### PR TITLE
docker: add `file`, adjust locale, and use python3 for ubuntu

### PR DESCRIPTION
**Pull request by scottwittenburg**
I've noticed that:

1. now spack chooses python3 by default, and 
2. one of the packages installed brings python3, but 
3. the boto3 module is not installed under python3, so that 
4. spack/ubuntu images no longer work for accessing S3 buckets

This change should make sure boto3 is installed under python3 as well as python, so that boto3 is available in either case.

### CODE REVIEW

**Comment by scottwittenburg** [comment]
@opadron The second commit here, adding locale support to the container, did indeed fix the issue I was seeing building `perl`:

```
==> [2019-10-31-16:38:52.377293] WRITE LOCK: /tmp/tmp.xZd3qHRjkt/spack/opt/spack/.spack-db/prefix_lock[9218058316734288330:1] [Released]
==> [2019-10-31-16:38:52.377378] WRITE LOCK: /tmp/root/spack-stage/.lock[6546459184478906453:1] [Released]
==> [2019-10-31-16:38:52.377429] WRITE LOCK: /tmp/root/spack-stage/.lock[203305486711886210:1] [Released]
==> [2019-10-31-16:38:52.386299] Error: UnicodeDecodeError: 'ascii' codec can't decode byte 0xe3 in position 3694: ordinal not in range(128)

/tmp/tmp.xZd3qHRjkt/spack/lib/spack/spack/package.py:1676, in build_process:
       1673                    echo = logger.echo
       1674                    self.log()
       1675
  >>   1676                # Run post install hooks before build stage is removed.
       1677                spack.hooks.post_install(self.spec)
       1678
       1679            # Stop timer.

See build log for details:
  /tmp/root/spack-stage/spack-stage-perl-5.30.0-77nd4zr7scrzl5yykxeoonx5uykverr2/spack-build-out.txt

Traceback (most recent call last):
  File "/tmp/tmp.xZd3qHRjkt/spack/lib/spack/spack/build_environment.py", line 797, in child_process
    return_value = function()
  File "/tmp/tmp.xZd3qHRjkt/spack/lib/spack/spack/package.py", line 1677, in build_process
    spack.hooks.post_install(self.spec)
  File "/tmp/tmp.xZd3qHRjkt/spack/lib/spack/spack/hooks/__init__.py", line 60, in __call__
    hook(*args, **kwargs)
  File "/tmp/tmp.xZd3qHRjkt/spack/lib/spack/spack/hooks/sbang.py", line 106, in post_install
    filter_shebangs_in_directory(directory, filenames)
  File "/tmp/tmp.xZd3qHRjkt/spack/lib/spack/spack/hooks/sbang.py", line 93, in filter_shebangs_in_directory
    filter_shebang(path)
  File "/tmp/tmp.xZd3qHRjkt/spack/lib/spack/spack/hooks/sbang.py", line 37, in filter_shebang
    original = original_file.read()
  File "/usr/lib/python3.6/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe3 in position 3694: ordinal not in range(128)
```

**Comment by scottwittenburg** [comment]
By the way, I really don't like the way the packages to be installed are organized into columns, it makes for unnecessarily large diffs resulting from really small changes.

Any votes for me re-organizing those into a single column in this PR?

**Comment by tgamblin** [comment]
Yes! Please PLEASE reorganize them!

File: share/spack/docker/ubuntu-1604.dockerfile [hunk 1]
```diff
@@ -17,16 +17,23 @@ COPY share $SPACK_ROOT/share
...
+RUN apt-get -yqq update                                        \
+ && apt-get -yqq install                                       \
+        build-essential ca-certificates curl       g++         \
+        gcc             gfortran        git        gnupg2      \
+        iproute2        lmod            locales    lua-posix   \
+        make            openssh-server  python     python-pip  \
+        python3-pip     tcl             unzip                  \
+ && locale-gen en_US.UTF-8                                     \
+ && pip install boto3                                          \
+ && pip3 install boto3                                         \
```
Comment (opadron) [inline share/spack/docker/ubuntu-1604.dockerfile]: I think we can drop python2 in favor of python3.  No need to have both in the same image.

File: share/spack/docker/ubuntu-1804.dockerfile [hunk 2]
```diff
@@ -17,16 +17,23 @@ COPY share $SPACK_ROOT/share
...
+RUN apt-get -yqq update                                        \
+ && apt-get -yqq install                                       \
+        build-essential ca-certificates curl       g++         \
+        gcc             gfortran        git        gnupg2      \
+        iproute2        lmod            locales    lua-posix   \
+        make            openssh-server  python     python-pip  \
+        python3-pip     tcl             unzip                  \
+ && locale-gen en_US.UTF-8                                     \
+ && pip install boto3                                          \
+ && pip3 install boto3                                         \
```
Comment (opadron) [inline share/spack/docker/ubuntu-1804.dockerfile]: rm python2 support

File: share/spack/docker/ubuntu-1604.dockerfile [hunk 3]
```diff
@@ -17,16 +17,23 @@ COPY share $SPACK_ROOT/share
...
  && rm -rf /var/lib/apt/lists/*
 
+# Add LANG default to en_US.UTF-8
+ENV LANGUAGE en_US.UTF-8
+ENV LANG en_US.UTF-8
+ENV LC_ALL en_US.UTF-8
```
Comment (opadron) [inline share/spack/docker/ubuntu-1604.dockerfile]: What is the locale situation on the centos images?  Perhaps we should also be taking these steps there?

File: share/spack/docker/ubuntu-1604.dockerfile (hunk 1 above)
Comment (scottwittenburg) [reply share/spack/docker/ubuntu-1604.dockerfile]: Done

File: share/spack/docker/ubuntu-1804.dockerfile (hunk 2 above)
Comment (scottwittenburg) [reply share/spack/docker/ubuntu-1804.dockerfile]: Done

File: share/spack/docker/ubuntu-1604.dockerfile (hunk 3 above)
Comment (scottwittenburg) [reply share/spack/docker/ubuntu-1604.dockerfile]: Perhaps, we could look it at some point.  But I'd rather not hold this PR up over it.

**Comment by scottwittenburg** [comment]
I've built this image locally and pushed it to my own repo on dockerhub, and I'm currently using it to run a bootstrapping pipeline on our cloud infrastructure.  That revealed another issue with the image (missing the `file` command needed for creating buildcaches), which I have fixed.  

I'll post back here if that pipeline runs successfully, and then maybe we can merge this.

**Comment by scottwittenburg** [comment]
@tgamblin This somehow reduced coverage by adding lines to Dockerfiles, and I'm not aware of anything we can do about that.

Unless you have objections, I'm thinking of merging this since it allows me to run pipelines on Ubuntu that work and can communicate to our S3 buckets.

**Pull request merged**